rules_daml/daml.bzl

```diff
@@ -1,94 +1,283 @@
+    )
+
+def _daml_build_impl(ctx):
+    name = ctx.label.name
+    daml_yaml = ctx.file.daml_yaml
+    damls = ctx.files.damls
+    dar_dict = ctx.attr.dar_dict
+    damlc = ctx.file._damlc
+    input_dars = [file_of_target(k) for k in dar_dict.keys()]
+    output_dar = ctx.outputs.dar
+    ctx.actions.run_shell(
+        tools = [damlc],
+        inputs = [daml_yaml] + damls + input_dars,
+        outputs = [output_dar],
+        progress_message = "Building DAML project %s" % name,
+        command = """
+            set -eou pipefail
+            tmpdir=$(mktemp -d)
+            trap "rm -rf $tmpdir" EXIT
+            cp -f {config} $tmpdir/daml.yaml
+            {cp_damls}
+            {cp_dars}
+            {damlc} build --project-root $tmpdir -o $PWD/{output_dar}
+        """.format(
+            config = daml_yaml.path,
+            cp_damls = "\n".join([
+                make_cp_command(
+                    src = daml.path,
+                    dest = "$tmpdir/" + daml.path,
+                )
+                for daml in damls
+            ]),
+            cp_dars = "\n".join([
+                make_cp_command(
+                    src = file_of_target(k).path,
+                    dest = "$tmpdir/" + v,
+                )
+                for k, v in dar_dict.items()
+            ]),
+            damlc = damlc.path,
+            output_dar = output_dar.path,
+        ),
+    )
+
+_daml_build = rule(
+    implementation = _daml_build_impl,
+    attrs = {
+        "daml_yaml": attr.label(
+            allow_single_file = True,
+            mandatory = True,
+            doc = "The daml.yaml config file.",
+        ),
+        "damls": attr.label_list(
             allow_files = [".daml"],
-            default = [],
-            doc = "Other DAML files that compilation depends on.",
+            mandatory = True,
+            doc = "DAML files in this DAML project.",
         ),
-        "target": attr.string(doc = "DAML-LF version to output"),
-        "damlc": attr.label(
-            executable = True,
-            cfg = "host",
-            allow_files = True,
-            default = Label("//compiler/damlc"),
+        "dar_dict": attr.label_keyed_string_dict(
+            mandatory = True,
+            doc = "Other DAML projects referenced by this DAML project.",
         ),
+        "dar": attr.output(
+            mandatory = True,
+            doc = "The generated DAR file.",
+        ),
+        "_damlc": _damlc,
     },
-    executable = False,
-    outputs = _daml_compile_outputs_impl,
 )
 
+def _extract_main_dalf_impl(ctx):
+    project_name = ctx.attr.project_name
+    project_version = ctx.attr.project_version
+    input_dar = ctx.file.dar
+    output_dalf = ctx.outputs.dalf
+    ctx.actions.run_shell(
+        inputs = [input_dar],
+        outputs = [output_dalf],
+        progress_message = "Extract DALF from DAR (%s)" % project_name,
+        command = """
+            set -eou pipefail
+            unzip -q {input_dar}
```

Comment:
Bazel offers `@bazel_tools//tools/zip:zipper` which you can use instead of `unzip`. See [here](https://github.com/bazelbuild/bazel/blob/18bd1056b00e1fd58638d4db81e02cfb7df8ce69/third_party/ijar/zip_main.cc#L361).

Reply:
@aherrmann-da  Thanks for this suggestion.
I changed from using `unzip` to `zipper x`. works great!
Unfortunately I think I may still have a problem because I am using `find`... is that available on windows?

(And a related question. Is `ln -s` possible on windows?)

Reply:
> (And a related question. Is `ln -s` possible on windows?)

No, symlinks on windows are pretty much unusable.

I believe you can get `find` from `msys2`. The other `dev_env_tool` rules in `WORKSPACE` are probably a useful reference.

Reply:
ok, I'm changed to getting `find` from the posix toolchain, as `scala.bzl` does
Lets see if it works...

Reply:
it did

Reply:
Ah right the posix toolchain is much nicer.